File: code/python/lists/leet_code_206.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverselist(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return None
        pre = None
        cur = head
        while cur.next is not None:
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt
        cur.next = pre
        return cur

    def recursive_reverse_list(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        rest_head = self.recursive_reverse_list(head.next)
        head.next.next = head
        head.next = None
        return rest_head

File: code/java/lists/src/main/java/org/swufe/datastructure/LeetCode206.java
package org.swufe.datastructure;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class LeetCode206 {
    public ListNode reverseList(ListNode head) {
        if (head == null) return null;
        ListNode pre = null;
        ListNode cur = head;
        while (cur.next != null) {
            ListNode next = cur.next;
            cur.next = pre;
            pre = cur;
            cur = next;
        }
        cur.next = pre;
        return cur;
    }

    public ListNode recursiveReverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode restHead = recursiveReverseList(head.next);
        head.next.next = head;
        head.next = null;
        return restHead;
    }
}
